materialize-slack/main.go

```diff
@@ -0,0 +1,357 @@
+					Type:   pm.Response_Validated_Constraint_FIELD_REQUIRED,
+					Reason: "The Slack materialization requires 'text'",
+				}
+			case projection.Field == "blocks":
+				constraint = pm.Response_Validated_Constraint{
+					Type:   pm.Response_Validated_Constraint_FIELD_REQUIRED,
+					Reason: "The Slack materialization would like 'blocks'",
+				}
+			case projection.IsPrimaryKey:
+				constraint = pm.Response_Validated_Constraint{
+					Type:   pm.Response_Validated_Constraint_LOCATION_REQUIRED,
+					Reason: "Keys are all required",
+				}
+			default:
+				constraint = pm.Response_Validated_Constraint{
+					Type:   pm.Response_Validated_Constraint_FIELD_OPTIONAL,
+					Reason: "All fields other than 'ts', 'text', and 'blocks' will be ignored",
+				}
+			}
+			constraints[projection.Field] = &constraint
+		}
+
+		if constraints["text"] == nil {
+			return nil, fmt.Errorf("'text' field required")
+		}
+
+		out = append(out, &pm.Response_Validated_Binding{
+			Constraints:  constraints,
+			DeltaUpdates: true,
+			ResourcePath: []string{res.Channel},
+		})
+	}
+
+	return &pm.Response_Validated{Bindings: out}, nil
+}
+
+func (driver) Apply(ctx context.Context, req *pm.Request_Apply) (*pm.Response_Applied, error) {
+	log.Debug("handling Apply request")
+	return &pm.Response_Applied{ActionDescription: "materialize-slack does not modify channels"}, nil
+}
+
+// Transactions implements the DriverServer interface.
+func (driver) NewTransactor(ctx context.Context, open pm.Request_Open) (pm.Transactor, *pm.Response_Opened, error) {
+	log.Debug("handling Transactions request")
+
+	var cfg config
+	if err := pf.UnmarshalStrict(open.Materialization.ConfigJson, &cfg); err != nil {
+		return nil, nil, fmt.Errorf("parsing endpoint config: %w", err)
+	}
+
+	var checkpoint driverCheckpoint
+	if err := pf.UnmarshalStrict(open.StateJson, &checkpoint); err != nil {
+		return nil, nil, fmt.Errorf("parsing driver checkpoint: %w", err)
+	}
+
+	api, err := cfg.buildAPI()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var bindings []*binding
+	for _, b := range open.Materialization.Bindings {
+		var res resource
+		err = json.Unmarshal(b.ResourceConfigJson, &res)
+		if err != nil {
+			return nil, nil, fmt.Errorf("unable to parse resource config: %w", err)
+		}
+		bindings = append(bindings, &binding{
+			resource:   res,
+			collection: string(b.Collection.Name),
+			spec:       b,
+		})
+	}
+
+	var transactor = &transactor{
+		api:      api,
+		bindings: bindings,
+	}
+
+	return transactor, &pm.Response_Opened{}, nil
+}
+
+type transactor struct {
+	api      *SlackAPI
+	bindings []*binding
+}
+
+type binding struct {
+	spec       *pf.MaterializationSpec_Binding
+	resource   resource
+	collection string
+}
+
+func buildDocument(b *binding, keys, values tuple.Tuple) map[string]interface{} {
+	var document = make(map[string]interface{})
+
+	// Add the keys to the document.
+	for i, value := range keys {
+		if i < len(b.spec.FieldSelection.Keys) {
```

Comment:
Since we are requiring that all keys be materialized I don't think this length check is required.